std-bits/table/src/main/java/org/enso/table/expressions/ExpressionVisitorImpl.java

```diff
@@ -0,0 +1,342 @@
+    public static final ThrowOnErrorListener INSTANCE = new ThrowOnErrorListener();
+
+    @Override
+    public void syntaxError(
+        Recognizer<?, ?> recognizer,
+        Object offendingSymbol,
+        int line,
+        int charPositionInLine,
+        String msg,
+        RecognitionException e)
+        throws SyntaxErrorException {
+      throw new SyntaxErrorException(msg, line, charPositionInLine);
+    }
+  }
+
+  public static class SyntaxErrorException extends RuntimeException {
+    private final int line;
+    private final int column;
+
+    public SyntaxErrorException(String message, int line, int column) {
+      super(message);
+      this.line = line;
+      this.column = column;
+    }
+
+    public int getLine() {
+      return line;
+    }
+
+    public int getColumn() {
+      return column;
+    }
+  }
+
+  public static Value evaluate(
+      String expression,
+      Function<String, Value> getColumn,
+      Function<Object, Value> makeConstantColumn,
+      String moduleName,
+      String typeName,
+      String[] variableArgumentFunctions)
+      throws UnsupportedOperationException, IllegalArgumentException {
+    var lexer = new ExpressionLexer(CharStreams.fromString(expression));
+    lexer.removeErrorListeners();
+    lexer.addErrorListener(ThrowOnErrorListener.INSTANCE);
+
+    var tokens = new CommonTokenStream(lexer);
+    var parser = new ExpressionParser(tokens);
+    parser.removeErrorListeners();
+    parser.addErrorListener(ThrowOnErrorListener.INSTANCE);
+
+    var visitor =
+        new ExpressionVisitorImpl(
+            getColumn, makeConstantColumn, moduleName, typeName, variableArgumentFunctions);
+
+    var expr = parser.prog();
+    return visitor.visit(expr);
+  }
+
+  private final Function<String, Value> getColumn;
+  private final Function<Object, Value> makeConstantColumn;
+  private final Function<String, Value> getMethod;
+  private final Set<String> variableArgumentFunctions;
+
+  private ExpressionVisitorImpl(
+      Function<String, Value> getColumn,
+      Function<Object, Value> makeConstantColumn,
+      String moduleName,
+      String typeName,
+      String[] variableArgumentFunctions) {
+    this.getColumn = getColumn;
+    this.makeConstantColumn = makeConstantColumn;
+
+    final Value module =
+        Context.getCurrent().getBindings("enso").invokeMember("get_module", moduleName);
+    final Value type = module.invokeMember("get_type", typeName);
+    this.getMethod = name -> module.invokeMember("get_method", type, name);
+
+    this.variableArgumentFunctions = new HashSet<>(Arrays.asList(variableArgumentFunctions));
+  }
+
+  private Value wrapAsColumn(Value value) {
+    if (value.isNull()) {
+      return makeConstantColumn.apply(value);
+    }
+
+    var metaObject = value.getMetaObject();
+    return metaObject != null && metaObject.asHostObject() instanceof Class<?>
+        ? makeConstantColumn.apply(value)
+        : value;
```

Comment:
I'm not sure if I understand this condition - I understand that if an object comes from Java, we pass it over to `makeConstantColumn`, but otherwise we just pass it as-is. Why is that so? The name `wrapAsColumn` is not clear enough to me to say what should be expected here.
Can we add some comment to this function explaining what it is meant to do?

Reply:
After looking forward at the code I think I now understand what is the meaning, but now I'm afraid of this logic.
It will make a constant column from any Java value but pass Enso values as-is. I guess for now it is OK, but if in the future we allow somehow in the expression to take hold of a singleton Enso value, this part of the logic will break - we'd expect it to wrap the value to a column, but it would be passed as-is instead.

Can we change the logic to detect the column? Maybe we can look at the MetaObject and see if it matches the `moduleName` (when Jaroslav's changes with MetaObjects for Atoms are integrated).

If not, let's at least add a comment warning that we may need to be careful in case of any changes.